### PR TITLE
add test for mix test --only async/sync filter

### PR DESCRIPTION
**Pull request by SteffenDE**
Relates to https://github.com/elixir-lang/elixir/pull/13639#issuecomment-2158497390, https://github.com/elixir-lang/elixir/commit/ecac839810905f2da826c5cca01cb4e0e9fc5a95

@josevalim

### CODE REVIEW

File: lib/ex_unit/test/ex_unit_test.exs [hunk 1]
```diff
@@ -1101,6 +1101,48 @@ defmodule ExUnitTest do
     assert third =~ "ThirdTestFIFO"
   end
 
+  for async? <- [true, false] do
```
Comment (josevalim) [inline lib/ex_unit/test/ex_unit_test.exs:1104]: Could we instead run the same modules twice, but with different arguments, in the same test? That should make it faster and simpler!

Comment (SteffenDE) [reply lib/ex_unit/test/ex_unit_test.exs]: yes we can, but I needed to adjust `run_with_filter`, see the new commit. I'm not sure if I like this more.

**Comment by josevalim** [comment]
Looks beautiful to me. If CI is happy, I am happy.

**Comment by josevalim** [comment]
:green_heart: :blue_heart: :purple_heart: :yellow_heart: :heart:

**Pull request merged**